src/components/MessageInput/hooks/useMessageInputState.ts

```diff
@@ -271,31 +282,35 @@ const messageInputReducer = <
     case 'clear':
       return makeEmptyMessageInputState();
 
-    case 'upsertAttachment': {
-      const attachmentIndex = state.attachments.findIndex(
-        (att) => att.$internal?.id && att.$internal?.id === action.attachment.$internal?.id,
-      );
-      const upsertedAttachment = mergeDeep(
-        state.attachments[attachmentIndex] ?? {},
-        action.attachment,
-      );
+    case 'upsertAttachments': {
       const attachments = [...state.attachments];
-      attachments.splice(attachmentIndex, 1, upsertedAttachment);
+      action.attachments.forEach((actionAttachment) => {
+        const attachmentIndex = state.attachments.findIndex(
+          (att) =>
+            att.localMetadata?.id && att.localMetadata?.id === actionAttachment.localMetadata?.id,
+        );
+
+        if (attachmentIndex === -1) {
+          attachments.push(actionAttachment);
+        } else {
+          const upsertedAttachment = mergeDeep(
```

Comment:
@arnautov-anton regarding your question about patching the attachments. It happens here.